include/sys/onoff.h

```diff
@@ -0,0 +1,533 @@
+ * providers or clients.  The state is to be initialized once, using
+ * onoff_service_init(), when the service provider is initialized.
+ * In case of error it may be reset through the
+ * onoff_service_reset() API.
+ */
+struct onoff_service {
+	/* List of clients waiting for completion of reset or
+	 * transition to on.
+	 */
+	sys_slist_t clients;
+
+	/* Function to invoke to transition the service to on. */
+	onoff_service_transition_fn start;
+
+	/* Function to invoke to transition the service to off. */
+	onoff_service_transition_fn stop;
+
+	/* Function to force the service state to reset, where
+	 * supported.
+	 */
+	onoff_service_transition_fn reset;
+
+	/* Mutex protection for flags, clients, releaser, and refs. */
+	struct k_spinlock lock;
+
+	/* Client to be informed when transition to off completes. */
+	struct onoff_client *releaser;
+
+	/* Flags identifying the service state. */
+	u16_t flags;
+
+	/* Number of active clients for the service. */
+	u16_t refs;
+};
+
+/**
+ * @brief Initialize an on-off service to off state.
+ *
+ * This function must be invoked exactly once per service instance,
+ * before any other on-off service API is invoked on the service.
+ *
+ * @param srv the service for which transition was requested.
+ *
+ * @param start the function used to (initiate a) transition from off
+ * to on.  This must not be null.  Include
+ * ONOFF_SERVICE_ISR_CALLABLE_START and/or
+ * ONOFF_SERVICE_SUSPENDABLE_START as appropriate in flags.
+ *
+ * @param stop the function used to (initiate a) transition from on to
+ * to off.  This must not be null.  Include
+ * ONOFF_SERVICE_ISR_CALLABLE_STOP and/or
+ * ONOFF_SERVICE_SUSPENDABLE_STOP as appropriate in flags.
+ *
+ * @param reset the function used to clear errors and force the
+ * service to an off state.  Pass null if the service cannot or need
+ * not be reset.  (Services where a transition operation can complete
+ * with an error notification should support the reset operation.)
+ * Include ONOFF_SERVICE_ISR_CALLABLE_RESET and/or
+ * ONOFF_SERVICE_SUSPENDABLE_RESET as appropriate in flags.
+ *
+ * @param flags any or all of the
+ * ONOFF_SERVICE_{ISR_CALLABLE,SUSPENDS}_{START,STOP,RESET} flags.
+ * Use of other flags produces an error.
+ *
+ * @retval 0 on success
+ * @retval -EINVAL if start, stop, or flags are invalid
+ */
+int onoff_service_init(struct onoff_service *srv,
+		       onoff_service_transition_fn start,
+		       onoff_service_transition_fn stop,
+		       onoff_service_transition_fn reset,
+		       u32_t flags);
+
+/**
+ * @brief Flag fields used to specify on-off client behavior.
+ *
+ * These flags control whether calls to onoff_service_request() and
+ * onoff_service_release() are synchronous or asynchronous, and for
+ * asynchronous operations how the operation result is communicated to
+ * the client.
+ */
+enum onoff_client_flags {
+	/* Indicates operations are to proceed only if the initiating
```

Comment:
It's intended to mean what `K_NO_WAIT` means for time-limited operations: if you can do it immediately fine, if not, fail.  This too needs to be updated for terminology changes in #20959.

Reply:
do you see a use case of such mode?

Reply:
Absolutely.  Many things that can be turned on and off will never delay (part of why "suspendable" is a service flag).  All of them will only delay on the first or last request.  If the caller does not expect a delay to be required the caller can allocate the client structure on the stack and be guaranteed a reference to it will not be held in the service.

It's this feature that makes it possible to request and release resources from ISRs, even if you still need to provide a way to re-issue the operation from thread context if it fails.